pkg/deps/meta.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deps
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	bundle "github.com/GoogleCloudPlatform/k8s-cluster-bundle/pkg/apis/bundle/v1alpha1"
+	"github.com/GoogleCloudPlatform/k8s-cluster-bundle/pkg/converter"
+	"github.com/blang/semver"
+)
+
+// A single requested dependency.
+type requestedDep struct {
+	componentName string
+	version       semver.Version
+}
+
+// String formats the requestedDep as a string
+func (r *requestedDep) String() string {
+	return fmt.Sprintf("{%s, %s}", r.componentName, r.version)
+}
+
+// depMeta contains dependency metadata about a single component.
+type depMeta struct {
+	componentName string
+	version       semver.Version
+	reqDeps       []*requestedDep
+	visibility    map[string]bool
+	annotations   map[string]string
+}
+
+// String prints a stringified version of the depMeta object.
+func (m *depMeta) String() string {
+	baseStr := fmt.Sprintf("{reference:{%s, %s}", m.componentName, m.version.String())
+	if len(m.reqDeps) > 0 {
+		baseStr = baseStr + fmt.Sprintf(" requirements:%v", m.reqDeps)
+	}
+	if len(m.visibility) > 0 {
+		baseStr = baseStr + fmt.Sprintf(" visibility:%v", m.visibility)
+	}
+	baseStr += "}"
+	return baseStr
+}
+
+// versionStr returns the version string.
+func (m *depMeta) versionStr() string { return m.version.String() }
+
+// ref makes a component reference.
+func (m *depMeta) ref() bundle.ComponentReference {
+	return bundle.ComponentReference{m.componentName, m.version.String()}
+}
+
+// visibleTo returns whether this component is visible to another component.
+func (m *depMeta) visibleTo(other *depMeta) bool {
+	if len(m.visibility) == 0 || m.visibility["@private"] {
```

Comment:
still not sold on the idea that we need public/private labels.  Still see this as complexity that we should avoid for now.